Feat/parse html from js template string

JeffersonCarvalh0 opened the pull request:
I was facing a very similar problem to the one related in https://github.com/lukasgeiter/gettext-extractor/issues/21 so I decided to give it a try. The idea is to create a JS extractor that accepts an HTML parser as a parameter, in order to parse the HTML from template strings (pretty much the same idea from the `embeddedJs` extractor).
Please, let me know how I can improve this PR, and thanks for this amazing library!

PS: I'm not sure the linter worked

Commit:
add htmlTemplate extractor

Commit:
add prepare script for local tests

Commit:
fix line number

Commit:
improve test

Commit:
remove dist files

lukasgeiter commented on .npmignore:
I don't think this file is needed

lukasgeiter commented:
Thanks for your PR!

I'm wondering if limiting this feature to only template string literals with no substitutions is maybe a bit too restrictive? What are your thoughts on that?

JeffersonCarvalh0 commented:
I agree with you @lukasgeiter , I'm just not sure about which types I could check for (there are barely any docs for the compiler API from typescript - if you found a helpful source somewhere, please I'd like to know)

I could test with other types I find fitting in the `.d.ts` files from typescript and add them to this feature

lukasgeiter commented:
No worries. I recommend you use a tool that makes it easy to explore the AST. Here's two I've used before:
- [astexplorer.net](https://astexplorer.net/)
- [ts-ast-viewer.com](https://ts-ast-viewer.com/)

For this feature there are two potential types of string literals to handle. First, regular string literals (double and single quotes). This one is quite straightforward: `SyntaxKind.StringLiteral`. Even better, use `isStringLiteralLike` to check for string literals and template strings without substitutions at once.

Secondly, there's the topic of template strings with substitutions. This one will be a bit more complex because you would first have to build a full string out of the parts that can then be passed to the html parser. Before you attempt that, I would be curious if you have this case in your code base. Just to get an idea if it's even worth it to implement support for template strings with substitutions.

JeffersonCarvalh0 commented:
Thanks for the tips!

As for html templates within template strings with substitutions, if I understood correctly, I actually have them in the code base, but they are already being extracted. My guess is that the JS extractor actually "sees" the JS code inside the template because it is already programmed to do that, regardless of the string surrounding it, so it doesn't matter if it is an HTML template (and the interpolated values don't generate more HTML code, they are literally calls to the i18n extraction function). I can't think about a use case where we'd have to write a specific logic to handle that

Commit:
improve the check in htmlTemplate extractor and add a new test for callExpression

JeffersonCarvalh0 commented:
I've added `isStringLiteralLike` as suggested and also added a new test case to illustrate the cases I have in my codebase where template substitutions are used (note that I've added this test to the `callExpression` tests, not the `htmlTemplate` tests)

lukasgeiter commented:
Yes, expressions inside template strings are JavaScript so they are properly extracted. There's no need for an explicit test for that. Otherwise I would have to add tests for every possible place a call expression can occur in JavaScript.

What I'm talking about is running the HTML parser/extractor on a template string with substitutions. For example:

```
`<div>
    <translate>Hello World</translate>
    ${substitution}
</div>`
```

lukasgeiter commented on tests/js/extractors/factories/htmlTemplate.test.ts:
```diff
@@ -0,0 +1,80 @@
+            const lineNumber = startLineFromNode ? lineNumberStart + startLineFromNode - 1 : lineNumberStart;
+            innerJsParser.parseString(textNode.value, fileName, { lineNumberStart: lineNumber });
+          }
+        }
```
Could you please stick to one of the built in HTML extractors for the tests?

Commit:
remove .npmignore

Commit:
update htmlTemplate test to use a built in html extractor

JeffersonCarvalh0 commented:
> No worries. I recommend you use a tool that makes it easy to explore the AST. Here's two I've used before:
> 
> * [astexplorer.net](https://astexplorer.net/)
> * [ts-ast-viewer.com](https://ts-ast-viewer.com/)
> 
> For this feature there are two potential types of string literals to handle. First, regular string literals (double and single quotes). This one is quite straightforward: `SyntaxKind.StringLiteral`. Even better, use `isStringLiteralLike` to check for string literals and template strings without substitutions at once.
> 
> Secondly, there's the topic of template strings with substitutions. This one will be a bit more complex because you would first have to build a full string out of the parts that can then be passed to the html parser. Before you attempt that, I would be curious if you have this case in your code base. Just to get an idea if it's even worth it to implement support for template strings with substitutions.

Sorry about the delay.

Now I got what you meant. I don't think I have this case in my codebase =/
I've just applied the changes you suggested

lukasgeiter commented on tests/js/extractors/factories/callExpression.test.ts:
```diff
@@ -128,6 +128,31 @@ describe('JS: Call Expression Extractor', () => {
                 ]);
             });
 
+            test('HTML inside a template string with substitution', () => {
```
This test doesn't really add anything, please revert this change.

lukasgeiter commented on tests/js/extractors/factories/htmlTemplate.test.ts:
```diff
@@ -0,0 +1,66 @@
+      #
+      #
+      #
+      #
```
What's up with the `#`? That's not even valid JS

JeffersonCarvalh0 replied:
Yeah, you're correct, I think I just had a brain fart while writing this one. My intention was to add line breaks to test if the parser would use the correct line numbers, but the `#` were not necessary indeed, simple line breaks would do the trick.

lukasgeiter commented on tests/js/extractors/factories/htmlTemplate.test.ts:
```diff
@@ -0,0 +1,66 @@
+      ]);
+    });
+
+    test('HTML inside a template literal', () => {
```
I'd like to see some variety of tests. Take a look at `tests/html/extractors/factories/embeddedJs.test.ts`, you should be able to write very similar tests for this case.

Additionally, it would be good to have a test for each of the supported type of strings (regular string literal & template string literal).

lukasgeiter replied:
If you try to implement similar tests to the ones from embeddedJs, I expect `with lineNumberStart option` to be failing with this extractor. To resolve that, we basically have to implement this change for the js parser: #59

JeffersonCarvalh0 replied:
I've added more tests, but I couldn't make the `separate line` one from `embeddedJs` work. I doesn't extract anything if I put `\n` inside the html

lukasgeiter replied:
The separate line case is already covered by the other tests you have written, so that's fine.

lukasgeiter reviewed (changes requested):
Thanks for the changes!
Let's not bother with templates that have substitutions then. Support can still be added later if requested.

Commit:
fix lineNumberStart and add more tests

lukasgeiter commented on src/js/parser.ts:
```diff
@@ -3,7 +3,7 @@ import * as ts from 'typescript';
 import { IMessage } from '../builder';
 
-export type IJsExtractorFunction = (node: ts.Node, sourceFile: ts.SourceFile, addMessage: IAddMessageCallback) => void;
+export type IJsExtractorFunction = (node: ts.Node, sourceFile: ts.SourceFile, addMessage: IAddMessageCallback, lineNumberStart?: number) => void;
```
Does `lineNumberStart` need to be optional? As far as I can tell, it should always be a number, no?

lukasgeiter commented on tests/js/extractors/factories/htmlTemplate.test.ts:
```diff
@@ -0,0 +1,148 @@
+    });
+
+    test('single line (template string)', () => {
+      jsParser.parseString("let itBe = \"<div> <translate> test </translate> </div>\"");
```
I'm liking the new tests! If you change the outer string to single quotes, you don't need to escape the double quotes: `'let itBe = "..."'`

lukasgeiter reviewed (changes requested):
Thanks for the changes, and apologies for my delay. I think we're almost there 🙂

Commit:
change lineNumberStart type and replace quotes on tests

JeffersonCarvalh0 commented:
Sorry for the delay on my part, these last weeks have been busy =/
Please let me know if there is anything else that needs to be changed

lukasgeiter commented:
Released with v3.8.0
